vnext/Microsoft.ReactNative/XamlLoadState.cpp

```diff
@@ -0,0 +1,178 @@
+
+#define NTDLL_FUNCTION(name) reinterpret_cast<decltype(&name)>(GetProcAddress(ntdll, #name))
+
+std::string XamlLoadState::GetXamlVersion(PCWSTR path) {
+  DWORD dwHandle;
+  DWORD size = GetFileVersionInfoSizeW(path, &dwHandle);
+  std::string buffer;
+  buffer.reserve(size);
+
+  if (GetFileVersionInfo(path, 0, size, const_cast<char *>(buffer.c_str()))) {
+    VS_FIXEDFILEINFO *fileInfo;
+    UINT len{};
+    if (VerQueryValueW(buffer.c_str(), L"\\", (void **)&fileInfo, &len)) {
+      const DWORD ls = fileInfo->dwFileVersionMS & 0x3fff;
+      DWORD ms = fileInfo->dwFileVersionMS >> 16;
+      // WinUI 3 Alpha has version 10.0.10011.16384. Convert 10.0 to 3.0
+      if (ms == 10) {
+        ms = 3;
+      }
+      std::stringstream v;
+      v << ms << "." << ls;
+      return v.str();
+    }
+  }
+
+  return {};
+}
+
+XamlLoadState::XamlLoadState() {
+  HMODULE mod{nullptr};
+
+  if (GetModuleHandle(wux.data())) {
+    RegisterDll(wux.data(), nullptr);
+  }
+  mod = GetModuleHandle(mux.data());
+  if (mod) {
+    wchar_t path[MAX_PATH]{};
+    if (GetModuleFileNameW(mod, path, ARRAYSIZE(path))) {
+      RegisterDll(mux.data(), path);
+    }
+  }
+
+  auto ntdll = GetModuleHandle(L"ntdll.dll");
+  auto pfn = NTDLL_FUNCTION(LdrRegisterDllNotification);
+  (*pfn)(0, MyNotificationFunction, nullptr, &m_cookie);
+}
+
+XamlLoadState::~XamlLoadState() {
+  auto ntdll = GetModuleHandle(L"ntdll.dll");
+  auto pfn = NTDLL_FUNCTION(LdrUnregisterDllNotification);
+  (*pfn)(m_cookie);
+}
+
+void XamlLoadState::RegisterDll(PCWSTR DllName, PCWSTR path) {
+  if (wux == DllName) {
+    switch (m_mode) {
+      case XamlDialect::Unknown:
+        m_mode = XamlDialect::SystemXAML;
+        return;
+      case XamlDialect::SystemXAML:
+        // should never get here
+        return;
+      case XamlDialect::WinUI:
+        throw std::exception("XAML dialect mismatch - WinUI 3 already loaded when attempting to load WUX");
+    }
+  } else if (mux == DllName) {
+    std::string newVersion = GetXamlVersion(path);
+    switch (m_mode) {
+      case XamlDialect::Unknown:
+        m_version = newVersion;
+        switch (m_version[0]) {
+          case '2':
+            m_mode = XamlDialect::SystemXAML;
+            break;
+          case '3':
+            m_mode = XamlDialect::WinUI;
+            break;
+          default:
+            throw std::exception("Unknown XAML version");
+            break;
+        }
+        break;
+
+      case XamlDialect::SystemXAML:
+        if (m_version == "" && newVersion[0] == '2') {
+          m_version = newVersion;
+        } else {
+          /* we've either:
+             -  loaded WUX and we're trying to load WinUI 3
+             -  loaded WinUI 2.x and we're trying to load WinUI 3 or a different WinUI 2.x
+             */
+          throw std::exception("XAML dialect mismatch");
+        }
+        break;
+
+      case XamlDialect::WinUI:
+        throw std::exception("WinUI 3 already loaded");
```

Comment:
>WinUI 3 [](start = 30, length = 7)

Again since this can face customers, maybe more explicit about the name:
```suggestion
Windows.UI.Xaml version 3
```

Reply:
@dannyvv the actual public facing name is WinUI 3